# Quiz 2: pandas with `elections.csv`

Write your answer in the empty cell under each question, run it, and check the output before typing it into Canvas.

In [1]:
import pandas as pd

elections = pd.read_csv('elections.csv')
elections

,Year,Candidate,Party,Popular vote,Result,%
0,1824,Andrew Jackson,Democratic-Republican,151271,loss,57.210122
1,1824,John Quincy Adams,Democratic-Republican,113142,win,42.789878
2,1828,Andrew Jackson,Democratic,642806,win,56.203927
3,1828,John Quincy Adams,National Republican,500897,loss,43.796073
4,1832,Andrew Jackson,Democratic,702735,win,54.574789
...,...,...,...,...,...,...
177,2016,Jill Stein,Green,1457226,loss,1.073699
178,2020,Joseph Biden,Democratic,81268924,win,51.311515
179,2020,Donald Trump,Republican,74216154,loss,46.858542
180,2020,Jo Jorgensen,Libertarian,1865724,loss,1.177979


## Question 1 (5 pts)

Write a line of pandas code that returns the `elections` table sorted in ascending order by `"%"`. Store your result in a variable named `sorted_df`.

In [2]:
# sort_values sorts the rows by the '%' column; ascending (smallest first) is the default.
sorted_df = elections.sort_values('%')
sorted_df

,Year,Candidate,Party,Popular vote,Result,%
156,2004,David Cobb,Green,119859,loss,0.098088
141,1992,Bo Gritz,Populist,106152,loss,0.101918
160,2004,Michael Peroutka,Constitution,143630,loss,0.117542
148,1996,John Hagelin,Natural Law,113670,loss,0.118219
165,2008,Cynthia McKinney,Green,161797,loss,0.123442
...,...,...,...,...,...,...
133,1984,Ronald Reagan,Republican,54455472,win,59.023326
79,1920,Warren Harding,Republican,16144093,win,60.574501
120,1972,Richard Nixon,Republican,47168710,win,60.907806
91,1936,Franklin Roosevelt,Democratic,27752648,win,60.978107


## Question 2 (2 pts)

Would calling `sorted_df.iloc[[0], :]` give the same result as `sorted_df.loc[[0], :]`? (Yes / No)

*Tip: run both and compare.*

**Answer: No**

In [3]:
# No. iloc picks by POSITION, loc picks by index LABEL.
# After sorting, the first row (position 0) is David Cobb, 2004 (label 156),
# while the row with label 0 is Andrew Jackson, 1824, which is somewhere else in sorted_df.
display(sorted_df.iloc[[0], :])   # first row by position
display(sorted_df.loc[[0], :])    # row whose index label is 0

,Year,Candidate,Party,Popular vote,Result,%
156,2004,David Cobb,Green,119859,loss,0.098088


,Year,Candidate,Party,Popular vote,Result,%
0,1824,Andrew Jackson,Democratic-Republican,151271,loss,57.210122


## Question 3 (3 pts)

Using Boolean slicing, write one line of pandas code that returns a DataFrame that only contains election results from the 2000s.

*"The 2000s" here means the year 2000 and later (2000–2020). If your teacher means only 2000–2009, use* `elections[(elections['Year'] >= 2000) & (elections['Year'] < 2010)]`*.*

In [4]:
# elections['Year'] >= 2000 gives True/False per row; elections[...] keeps the True rows.
elections[elections['Year'] >= 2000]

,Year,Candidate,Party,Popular vote,Result,%
151,2000,Al Gore,Democratic,50999897,loss,48.491813
152,2000,George W. Bush,Republican,50456002,win,47.974666
153,2000,Harry Browne,Libertarian,384431,loss,0.365525
154,2000,Pat Buchanan,Reform,448895,loss,0.426819
155,2000,Ralph Nader,Green,2882955,loss,2.741176
156,2004,David Cobb,Green,119859,loss,0.098088
157,2004,George W. Bush,Republican,62040610,win,50.771824
158,2004,John Kerry,Democratic,59028444,loss,48.306775
159,2004,Michael Badnarik,Libertarian,397265,loss,0.325108
160,2004,Michael Peroutka,Constitution,143630,loss,0.117542


## Question 4 (5 pts)

Write one line of pandas code that returns a Series, where the index is the `"Party"` and the values are how many times that party won an election.

In [5]:
# .loc[rows, column] picks the winning rows and the Party column in one step,
# then value_counts() counts how often each party appears.
elections.loc[elections["Result"] == "win", "Party"].value_counts()

Party
Democratic               23
Republican               23
Whig                      2
Democratic-Republican     1
National Union            1
Name: count, dtype: int64

## Question 5 (5 pts)

Write one line of pandas code that returns a Series whose index is the years and whose values are the number of candidates that participated in those years' elections.

In [6]:
# groupby('Year') puts all candidates of the same year together; .size() counts the rows in each group.
elections.groupby('Year').size()

Year
1824    2
1828    2
1832    3
1836    3
1840    2
1844    2
1848    3
1852    3
1856    3
1860    4
1864    2
1868    2
1872    2
1876    2
1880    3
1884    4
1888    4
1892    4
1896    4
1900    3
1904    5
1908    4
1912    5
1916    4
1920    5
1924    3
1928    3
1932    4
1936    4
1940    3
1944    2
1948    6
1952    3
1956    3
1960    2
1964    2
1968    3
1972    3
1976    6
1980    5
1984    3
1988    4
1992    5
1996    7
2000    5
2004    6
2008    6
2012    4
2016    6
2020    4
dtype: int64

## Question 6 (5 pts)

Write a line of pandas code that creates a filtered DataFrame named `filtered_parties` from the `elections` dataset and keeps only the parties that have at least one election `%` more than 50%.

In [7]:
# groupby("Party") makes one group per party. .filter(...) keeps a WHOLE group (all its rows)
# only if the function returns True for it. lambda group: ... gets one party's rows;
# (group["%"] > 50) is True/False per election, and .any() is True if at least one is True.
filtered_parties = elections.groupby("Party").filter(lambda group: (group["%"] > 50).any())
filtered_parties

,Year,Candidate,Party,Popular vote,Result,%
0,1824,Andrew Jackson,Democratic-Republican,151271,loss,57.210122
1,1824,John Quincy Adams,Democratic-Republican,113142,win,42.789878
2,1828,Andrew Jackson,Democratic,642806,win,56.203927
4,1832,Andrew Jackson,Democratic,702735,win,54.574789
7,1836,Hugh Lawson White,Whig,146109,loss,10.005985
...,...,...,...,...,...,...
171,2012,Mitt Romney,Republican,60933504,loss,47.384076
173,2016,Donald Trump,Republican,62984828,win,46.407862
176,2016,Hillary Clinton,Democratic,65853514,loss,48.521539
178,2020,Joseph Biden,Democratic,81268924,win,51.311515


## Question 7 (8 pts)

Write a line of pandas code that uses the `filtered_parties` DataFrame to return a new DataFrame with row indices that correspond to the year and columns that correspond to each party. Each entry should be the total percentage of votes for all the candidates that ran during that particular year for the specified party. Missing values (a party with no candidate that year) should be `0`.

Expected first rows:

| Year | Democratic | Democratic-Republican | National Union | Republican | Whig |
|---|---|---|---|---|---|
| 1824 | 0.000000 | 100 | 0.000000 | 0.000000 | 0.000000 |
| 1828 | 56.203927 | 0 | 0.000000 | 0.000000 | 0.000000 |

In [8]:
# Same pivot_table pattern as in the lab (3d), but adding up '%' instead of counting:
#   index='Year'      -> one row per year
#   columns='Party'   -> one column per party
#   values='%'        -> the numbers to combine
#   aggfunc='sum'     -> add up the % of all of that party's candidates that year
#   fill_value=0      -> 0 instead of NaN when a party had no candidate that year
filtered_parties.pivot_table(index='Year', columns='Party', values='%', aggfunc='sum', fill_value=0)

Party,Democratic,Democratic-Republican,National Union,Republican,Whig
Year,,,,,
1824,0.000000,100.0,0.000000,0.000000,0.000000
1828,56.203927,0.0,0.000000,0.000000,0.000000
1832,54.574789,0.0,0.000000,0.000000,0.000000
1836,52.272472,0.0,0.000000,0.000000,47.727528
1840,46.948787,0.0,0.000000,0.000000,53.051213
1844,50.749477,0.0,0.000000,0.000000,49.250523
1848,42.552229,0.0,0.000000,0.000000,47.309296
1852,51.013168,0.0,0.000000,0.000000,44.056548
1856,45.306080,0.0,0.000000,33.139919,0.000000
